### PR TITLE
[RLlib] SlateQ training iteration function.

### PR DESCRIPTION
**Pull request by sven1977**
SlateQ training iteration function.
* Set `_disable_execution_plan_api=True` by default for SlateQ.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/slateq/slateq_tf_policy.py [hunk 1]
```diff
@@ -157,21 +157,18 @@ def build_slateq_losses(
     q_clicked = tf.gather(replay_click_q, clicked_indices)
     target_clicked = tf.gather(target, clicked_indices)
 
-    def get_train_op():
-        td_error = q_clicked - target_clicked
-        if policy.config["use_huber"]:
-            loss = huber_loss(td_error, delta=policy.config["huber_threshold"])
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_tf_policy.py:163]: did we get rid of the option to use these huber losses? Why remove all of this, and replace with only one option for the loss? Just curious, but otherwise lgtm

File: rllib/agents/slateq/slateq_tf_policy.py [hunk 2]
```diff
@@ -157,21 +157,18 @@ def build_slateq_losses(
...
-        pred=tf.greater(tf.reduce_sum(input_tensor=clicked), 0),
-        true_fn=get_train_op,
-        false_fn=lambda: (tf.constant(0.0), tf.constant(0.0)),
-        name="",
+    td_error = tf.where(
+        tf.cast(clicked, tf.bool),
+        replay_click_q - target,
+        tf.zeros_like(train_batch[SampleBatch.REWARDS]),
     )
+    if policy.config["use_huber"]:
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_tf_policy.py:165]: oh wait whoops i totally missed this my bad

File: rllib/agents/slateq/slateq_tf_policy.py [hunk 3]
```diff
@@ -206,7 +204,8 @@ def build_slateq_stats(policy: Policy, batch) -> Dict[str, TensorType]:
         "next_q_target_slate": policy._next_q_target_slate,
         "next_q_target_max": policy._next_q_target_max,
         "target_clicked": policy._target_clicked,
-        "td_error": policy._mean_td_error,
+        "td_error": policy._td_error,
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_tf_policy.py:207]: if td error is a tensor, how is this supposed to be logged as opposed to mean_td_error which appears to be a scalar?

File: rllib/agents/slateq/slateq_torch_policy.py [hunk 4]
```diff
@@ -198,7 +197,8 @@ def get_train_op():
     model.tower_stats["next_q_target_max"] = torch.mean(next_q_target_max)
     model.tower_stats["target_clicked"] = torch.mean(target_clicked)
     model.tower_stats["q_loss"] = loss
-    model.tower_stats["mean_td_error"] = torch.mean(mean_td_error)
+    model.tower_stats["td_error"] = td_error
+    model.tower_stats["mean_td_error"] = mean_td_error
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_torch_policy.py:201]: pretty much same questions here

**Review by avnishn — COMMENTED** [review]
I don't see a training iteration function in the diff here -- maybe github is off, but maybe you forgot to upload a commit? LMK

File: rllib/agents/slateq/slateq_tf_policy.py (hunk 3 above)
Comment (sven1977) [reply rllib/agents/slateq/slateq_tf_policy.py:207]: `td_error` needs to have the same batch dim as the row-indices from the buffer, so we know, which rows to update priorities for.

`mean_td_error` is "only" used for reporting.

Comment (sven1977) [reply rllib/agents/slateq/slateq_tf_policy.py:207]: In other words: All off-policy algos with a prio. replay buffer must log this stats from their `stats_fn`.

File: rllib/agents/slateq/slateq_tf_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/slateq/slateq_tf_policy.py:165]: :)

**Comment by sven1977** [comment]
@avnishn , we use the exact same `training_iteration` function as DQN now. That's why I had to add the `td_error` stats. Makes SlateQ a little more powerful and the code base simpler.

File: rllib/agents/dqn/dqn.py [hunk 5]
```diff
@@ -202,6 +202,90 @@ def get_default_policy_class(
...
+                self.workers.local_worker().foreach_policy_to_train(
+                    lambda p, pid: pid in to_update and p.update_target()
+                )
+                self._counters[NUM_TARGET_UPDATES] += 1
+                self._counters[LAST_TARGET_UPDATE_TS] = cur_ts
+
+            # Update weights and global_vars - after learning on the local worker -
+            # on all remote workers.
+            global_vars = {
+                "timestep": self._counters[NUM_ENV_STEPS_SAMPLED],
```
Comment (avnishn) [inline rllib/agents/dqn/dqn.py:281]: If this is using LR schedule or Entropy schedule, shouldn't this be num agent steps trained?

File: rllib/agents/dqn/simple_q.py [hunk 6]
```diff
@@ -258,10 +258,13 @@ def training_iteration(self) -> ResultDict:
...
             self._counters[LAST_TARGET_UPDATE_TS] = cur_ts
 
-        # Update remote workers' weights after learning on local worker.
-        if self.workers.remote_workers():
-            with self._timers[SYNCH_WORKER_WEIGHTS_TIMER]:
-                self.workers.sync_weights()
+        # Update weights and global_vars - after learning on the local worker - on all
+        # remote workers.
+        global_vars = {
+            "timestep": self._counters[NUM_ENV_STEPS_SAMPLED],
```
Comment (avnishn) [inline rllib/agents/dqn/simple_q.py:264]: same comment here as above

File: rllib/agents/slateq/slateq.py [hunk 7]
```diff
@@ -78,16 +79,20 @@
     "huber_threshold": 1.0,
 
     # === Replay buffer ===
-    # Size of the replay buffer. Note that if async_updates is set, then
-    # each worker will have a replay buffer of this size.
-    "buffer_size": DEPRECATED_VALUE,
     "replay_buffer_config": {
-        "type": "MultiAgentReplayBuffer",
+        # Enable the new ReplayBuffer API.
+        "_enable_replay_buffer_api": True,
```
Comment (avnishn) [inline rllib/agents/slateq/slateq.py:84]: is there a reason that you have to set this and its not just the case already? 

Does this make Slateq use another replay buffer api?

File: rllib/agents/slateq/slateq_tf_policy.py (hunk 3 above)
Comment (avnishn) [reply rllib/agents/slateq/slateq_tf_policy.py:207]: oh gotcha -- That makes sense.

**Review by avnishn — APPROVED** [review]
This all looks pretty good to me except the global vars that we would use for updating rollout workers -- I guess since sampling is synchronous, my comments don't really matter, so Imma go ahead and approve.

File: rllib/agents/dqn/dqn.py (hunk 5 above)
Comment (sven1977) [reply rllib/agents/dqn/dqn.py:281]: So I would leave it for now to reflect our current behavior.
Definitely worth looking into this and maybe provide a better per-policy fix for this.

File: rllib/agents/slateq/slateq.py (hunk 7 above)
Comment (sven1977) [reply rllib/agents/slateq/slateq.py:84]: Yeah, it'll use the new replay buffer API we are currently rolling out across RLlib.
